### ISSUE TITLE
ActionIfExists.APPEND and ActionIfExists.REPLACE switched?

### ISSUE THREAD

**Issue by pgrond** [LeMyst/WikibaseIntegrator#363]
I think the two are switched in the `claims.add()` method: https://github.com/LeMyst/WikibaseIntegrator/blob/9cc5523577beadec73e1cf594514a5d7d93f976e/wikibaseintegrator/models/claims.py#L80

Shouldn't force update be applied if the `action_if_exists == ActionIfExists.REPLACE`?

**Comment by LeMyst** [comment]
I think there's a misunderstand between KEEP, FORCE_APPEND, APPEND and REPLACE :

For an entity with already a P1 claim and you want another P1 claim.

- KEEP: The original claim will be kept and the new one will not be added (because there is already one with this property number)
- APPEND: The new claim will be added only if the new one is different (by comparing values)
- FORCE_APPEND: The new claim will be added even if already exists
- REPLACE: The new claim will replace the old one

Do you have a code example?

**Comment by dpriskorn** [comment]
I suggest we add that to the readme 😀

**Comment by pgrond** [comment]
Yes, that is exactly how I expect the switches for the ENUM to behave, but the force logic is inside the APPEND clause, not inside the REPLACE clause. See https://github.com/LeMyst/WikibaseIntegrator/blob/9cc5523577beadec73e1cf594514a5d7d93f976e/wikibaseintegrator/models/claims.py#L84

**Comment by pgrond** [comment]
Hmm, I still think something is off. `self.claims[property]` is a `list[Claim]` right? So the `.update()` on line 85 is referring to the update method of the `Claim` class, that overwrites all the properties of the existing claim:
```
    def update(self, claim: Claim) -> None:
        self.mainsnak = claim.mainsnak
        self.qualifiers = claim.qualifiers
        self.qualifiers_order = claim.qualifiers_order
        self.rank = claim.rank
        self.references = claim.references
```
So "force" seems like a good wording, it replaces the data that is already there for an existing claim.

On the other hand, the code that says to replace the old claim:
```
elif action_if_exists == ActionIfExists.REPLACE:
    if claim not in self.claims[property]:
        self.claims[property].append(claim)
```
only adds the claim to the list of claims if it is not already present.

**Comment by LeMyst** [comment]
Ok, I found "the trick", it's just above:
https://github.com/LeMyst/WikibaseIntegrator/blob/e1e602d02383404939d0eca80f5eee0eb2f413ca/wikibaseintegrator/models/claims.py#L58-L68

At this point, if I update the statements of property P31, all existing statements are marked as removed, except if the statements is already is our claims that we want to add.

Later, at line 89-91, we add all the missing claims.

To summarize:
1. We removed everything we don't want to add
2. We add everything not already present

I think I can remove the "TODO:"

It's clear for me now, do you see any flaw?

**Comment by LeMyst** [comment]
> So "force" seems like a good wording, it replaces the data that is already there for an existing claim.

Yeah, I didn't read the code carefully.

**Comment by pgrond** [comment]
Ah, now I get it (after reading the tests ;)). APPEND will add or **overwrite** an existing claim, REPLACE will remove **all** existing claims on that property and replace it with new ones.

So the code seems correct for this behaviour, but the wording threw me off.  If I add 1 claim with `action_if_exists=ActionIfExists.REPLACE`, I would not expect that all my existing claims would be removed, but that the values for an existing claim would be replaced. So the behaviour in the APPEND clause.

Maybe rename the constants to APPEND_OR_REPLACE and REPLACE_ALL? To make the behaviour a bit more clear?

**Comment by LeMyst** [comment]
Changed in 4aa6133847920b45fcc3abde804ab3a0e8b2786f

Thanks
